lotman/lotman_linux.go

```diff
@@ -314,9 +333,580 @@ func getFederationIssuer() (string, error) {
+			}
+			for key, value := range policyLotMap {
+				lotMap[key] = value
+			}
+		}
+	} else {
+		lotMap = policyLotMap
+	}
+
+	cacheDisks := param.Cache_DataLocations.GetStringSlice()
+	log.Tracef("Cache data locations being tracked by Lotman: %v", cacheDisks)
+	var totalDiskSpaceB uint64
+	for _, disk := range cacheDisks {
+		diskSpace, _, err := getDiskUsage(disk)
+		if err != nil {
+			return internalLots, errors.Wrapf(err, "error getting disk usage for filesystem path %s", disk)
+		}
+		totalDiskSpaceB += diskSpace
+	}
+
+	// Now guarantee our special "default" and "root" lots if the user hasn't provided them
+	// in the config. For now, these lots must always exist because they're used to make sure
+	// all data is tied to a lot (default) and that the requirement of a root lot is satisfied
+	// without allowing discovered lots to gain rootly status.
+	federationIssuer, err := getFederationIssuer()
+	if err != nil {
+		return internalLots, errors.Wrap(err, "Unable to determine federation issuer which is needed by Lotman to determine lot ownership")
+	}
+	if federationIssuer == "" {
+		return internalLots, errors.New("The detected federation issuer, which is needed by Lotman to determine lot/namespace ownership, is empty")
+	}
+	rootDedGB := bytesToGigabytes(totalDiskSpaceB)
+	zero := float64(0)
+	if _, exists := lotMap["default"]; !exists {
+		lotMap["default"] = Lot{
+			LotName: "default",
+			Owner:   federationIssuer,
+			Parents: []string{"default"},
+			MPA: &MPA{
+				// Set default values to 0 and let potential reallocation happen later.
+				DedicatedGB:     &zero,
+				OpportunisticGB: &zero,
+				MaxNumObjects:   &Int64FromFloat{Value: 0}, // Purge plugin doesn't yet use this, set to 0.
+			},
+
+		}
+	}
+	if _, exists := lotMap["root"]; !exists {
+		lotMap["root"] = Lot{
+			LotName: "root",
+			Owner:   federationIssuer,
+			Parents: []string{"root"},
+
+			Paths: []LotPath{
+				{
+					Path: 	"/",
+					Recursive: false, // setting this to true would prevent any other lot from claiming a path
+				},
+			},
+			MPA: &MPA{
+				// Max out dedicatedGB so the root lot never purges. All other lots should be tied to their own policies.
+				DedicatedGB: &rootDedGB,
+				OpportunisticGB: &zero,
+				MaxNumObjects:   &Int64FromFloat{Value: 0}, // Purge plugin doesn't yet use this, set to 0.
+			},
+		}
+	}
+
+	log.Tracef("Lotman will split lot disk space quotas amongst the discovered disk space: %vB", totalDiskSpaceB)
+	if policy.DivideUnallocated {
+		log.Traceln("Dividing unallocated space among lots")
+		divideRemainingSpace(&lotMap, totalDiskSpaceB)
+	}
+
+	// Set up lot timestamps (creation, expiration, deletion) if needed
+	configLotTimestamps(&lotMap)
+
+	internalLots, err = topoSort(lotMap)
+	if err != nil {
+		return internalLots, errors.Wrap(err, "error sorting lots prior to instantiation")
+	}
+
+	log.Tracef("Internal lot configuration: %+v", internalLots)
+	err = validateLotsConfig(internalLots, totalDiskSpaceB)
```

Comment:
Is there any reason why this validation of lots config is happening after the topoSort? Since this check is faster O(N) versus O(N+M), it should be run first to avoid wasting time on a more expensive check in case this fails.